### PR TITLE
add wrapper to vec_basic

### PR DESCRIPTION
**Pull request by irexyc**
Add some wrapper for vec_basic, hope the implementation is appropriate.

some test code
```
x <- S('x')
y <- S('y')
vec <- symengine:::vecbasic_new()
vec <- symengine:::vecbasic_push_back(vec, x)
vec <- symengine:::vecbasic_push_back(vec, y)
print(vec)
length(vec)
symengine:::vecbasic_get(vec,1)
z <- S('z')
vec <- symengine:::vecbasic_set(vec, 2, z)
symengine:::vecbasic_get(vec,2)
vec <- symengine:::vecbasic_erase(vec,1)
symengine:::vecbasic_get(vec,1)

x <- S('2')
y <- S('1')
z <- S('9')
vec <- symengine:::vecbasic_new()
vec <- symengine:::vecbasic_push_back(vec, x)
vec <- symengine:::vecbasic_push_back(vec, y)
vec <- symengine:::vecbasic_push_back(vec, z)
max(vec)
min(vec
```

### CODE REVIEW

**Comment by certik** [comment]
@Marlin-Na would you mind reviewing this please?

**Comment by Marlin-Na** [comment]
Hi @irexyc ,

Again, thanks a lot for this pull request. This is a good start to provide the support of `vecbasic` in R.
However, I am not sure whether it is necessary to give R level access to low-level C functions `vecbasic_push_back`, `vecbasic_set`, `vecbasic_earse`. These functions modify the R object in place, which violates the copy-on-modify semantics. For example:
```R
vec <- symengine:::vecbasic_new()
symengine:::vecbasic_push_back(vec, S("x"))
vec2 <- vec
symengine:::vecbasic_push_back(vec, S("y"))
# (Symbol)	x
# (Symbol)	y
vec2
# (Symbol)	x
# (Symbol)	y
```
As shown above, both vec and vec2 are modified.

I think what we need to do is to support `vecbasic` is to provide the following functions:
- `vecbasic_concentrate`: concentrate two `vecbasic` or `basic` objects, working like `c` in R.
- `vecbasic_subset`: can be used for `[`, accepts a `vecbasic` and an index, returns a `vecbasic`
- `vecbasic_get`: can be used for `[[`, accepts a `vecbasic` and an integer, returns a `basic`
- `vecbasic_assign` can be used for `[<-`, accepts two `vecbasic`, an index, returns a `vecbasic`

These functions should be vectorized and should be implemented in C level.

You can keep this pull request open. I may try to modify your version and push to your branch.

Best regards

File: src/vector.cpp [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+SEXP vecbasic_concentrate(SEXP dots) {
+    R_len_t len = Rf_length(dots);
+    SEXP    out = PROTECT(sexp_vecbasic());
+    for (R_xlen_t i = 0; i < len; i++) {
+        if (R_compute_identical(R_ExternalPtrTag(VECTOR_ELT(dots, i)), Rf_mkString("basic_struct*"), 15)) {
+            sexp_vecbasic_push_back(out, VECTOR_ELT(dots, i));
+        } else if (R_compute_identical(R_ExternalPtrTag(VECTOR_ELT(dots, i)), Rf_mkString("CVecBasic*"), 15)) {
+            vecbasic_concentrate_vec(out, VECTOR_ELT(dots, i));
+        } else {
+            Rf_error("The %dth object's type is error'", i);
```
Comment (Marlin-Na) [inline src/vector.cpp]: It should `i+1`, I think.

**Comment by Marlin-Na** [comment]
Hi @irexyc , thanks again for your pull request. I am going to merge this.

**Pull request merged**